### PR TITLE
Remove unnecessary files after hexo generate

### PR DESCRIPTION
**Pull request by 1v9**

### CODE REVIEW

**Comment by stevenjoezhang** [comment]
In fact, there may be a lot of unnecessary files in source/lib, but they are hard to remove...

**Comment by jiangtj** [comment]
If cdn setted, may need remove...

**Comment by 1v9** [comment]
Only touch those files belong to NexT, not plugins installed to `source/lib`.

**Comment by 1v9** [comment]
I have found a good method, wait... 🚀

**Comment by 1v9** [comment]
Sorry guys it's truly difficult for me 😂.

Hexo has [`config.ignore`](https://github.com/hexojs/hexo/blob/f172774d37f8dffe01d9acecd3613a8b6782e22a/lib/box/index.js#L32) which could perfectly ignore directories, it's different from `config.exclude`  and `config.skip_render`. You know we merge config on `generateBefore` and hexo also has built-in merge config functions so it may be complicated.

Then I tried to operate files by `hexo-fs` on a lot of filters or events, failed again. Only `deployBefore` is easy to configure but too restrictive. Need to find a good way to handle `hexo g`, `hexo s`, `hexo d`.

`hexo.route` used by this PR only supports [`string`](https://github.com/hexojs/hexo/blob/1f1275738a7b588359dc0263a31143233a4f5993/lib/hexo/router.js#L91) so I listed all files.

Both `hexo.route` and `config.ignore` solve the problem, but one is not very beautiful and the other is a bit complicated.

**Comment by jiangtj** [comment]
> Both `hexo.route` and `config.ignore` solve the problem, but one is not very beautiful and the other is a bit complicated.

`hexo.route` is enough, it can remove the file with the specified path before the file is generated.

> `hexo.route` used by this PR only supports [`string`](https://github.com/hexojs/hexo/blob/1f1275738a7b588359dc0263a31143233a4f5993/lib/hexo/router.js#L91) so I listed all files.

Path can not be a String?

> Hexo has [`config.ignore`](https://github.com/hexojs/hexo/blob/f172774d37f8dffe01d9acecd3613a8b6782e22a/lib/box/index.js#L32) which could perfectly ignore directorie

Are you sure that we can get the configuration of the theme before hexo calls `config.ignore`?

**Comment by 1v9** [comment]
@jiangtj it's long 😂

```bash
$ de --debug
WARN  [Hexo-Filters]---------------------------------after_init starts
WARN  [Hexo-Filters]---------------------------------after_clean starts
INFO  Deleted public folder.
WARN  [Hexo-Filters]---------------------------------before_exit starts
WARN  [Hexo-Events]---------------------------------exit starts        
03:12:20.604 DEBUG Writing database to C:\Users\mkb\1v9/db.json
03:12:20.629 DEBUG Hexo version: 3.9.0
03:12:20.636 DEBUG Working directory: ~\1v9\
03:12:20.883 DEBUG Config loaded: ~\1v9\_config.yml
03:12:20.968 DEBUG Plugin loaded: ...
...
03:12:21.818 DEBUG Script loaded: ...
...
03:12:23.077 WARN  [Hexo-Filters]---------------------------------after_init starts
03:12:23.079 DEBUG Loading database.
03:12:23.086 INFO  Start processing
03:12:23.248 DEBUG Processed: ...
...
03:12:24.681 DEBUG Theme config loaded.
03:12:24.683 DEBUG Processed: _config.yml
...
03:12:30.488 WARN  [Hexo-Events]---------------------------------generateBefore starts
03:12:30.623 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.639 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.640 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.641 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.742 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.846 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.914 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.963 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.964 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.965 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.965 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.966 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.967 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.968 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.969 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.969 WARN  [Hexo-Filters]---------------------------------before_post_render starts
03:12:30.972 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\Debian 安装 HAProxy.md
03:12:30.978 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\NexT 添加今日诗词.md
03:12:31.044 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\acme.sh 与 letsencrypt.md
03:12:31.160 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\使用 Gulp 精简 Hexo.md
03:12:31.173 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\pixiv 漫画下载脚本.md
03:12:31.177 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\备份文件至 Github 私有仓库.md
03:12:31.253 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\编译安装 NGINX.md
03:12:31.255 DEBUG Rendering post: C:\Users\mkb\1v9\source\_posts\阻止针对 NGINX 的恶意访问.md
03:12:32.686 DEBUG Generator: page
03:12:32.697 DEBUG Generator: post
03:12:32.704 DEBUG Generator: archive
03:12:32.704 DEBUG Generator: feed
03:12:32.706 DEBUG Generator: category
03:12:32.707 DEBUG Generator: xml
03:12:32.711 DEBUG Generator: index
03:12:32.762 DEBUG Generator: leancloud_counter_security_generator
03:12:32.763 DEBUG Generator: tag
03:12:32.765 DEBUG Generator: asset
03:12:32.792 WARN  [Hexo-Events]---------------------------------generateAfter starts
03:12:32.848 WARN  [Hexo-Filters]---------------------------------after_generate starts
03:12:32.878 INFO  Files loaded in 9.8 s
03:12:32.948 DEBUG Rendering HTML page: categories/index.html
03:12:33.103 DEBUG Rendering HTML page: favorites/index.html
03:12:33.185 DEBUG Rendering HTML page: photos/index.html
03:12:33.250 DEBUG Rendering HTML page: tags/index.html
03:12:33.316 DEBUG Rendering HTML post: post/add-today-poetry-for-theme-next.html
03:12:33.435 DEBUG Rendering HTML post: post/pixiv-manga-download-script.html
03:12:33.540 DEBUG Rendering HTML post: post/compiling-and-installing-nginx.html
03:12:33.664 DEBUG Rendering HTML post: post/block-malicious-access-to-nginx.html
03:12:33.834 DEBUG Rendering HTML post: post/use-gulp-to-minify-hexo.html
03:12:34.168 DEBUG Rendering HTML post: post/backup-files-to-github-private-repo.html
03:12:34.455 DEBUG Rendering HTML post: post/acme.sh-and-letsencrypt.html
03:12:34.814 DEBUG Rendering HTML post: post/debian-install-haproxy.html
03:12:35.090 DEBUG Rendering HTML archive: archives/2019/index.html
03:12:35.288 DEBUG Rendering HTML archive: archives/index.html
03:12:35.399 DEBUG Rendering HTML archive: archives/2019/01/index.html
03:12:35.491 DEBUG Rendering HTML archive: archives/2019/02/index.html
03:12:35.571 DEBUG Rendering HTML archive: archives/2019/03/index.html
03:12:35.661 DEBUG Rendering HTML category: categories/HAProxy/index.html
03:12:35.748 DEBUG Rendering HTML category: categories/NexT/index.html
03:12:35.805 DEBUG Rendering HTML category: categories/HTTPS/index.html
03:12:35.869 DEBUG Rendering HTML category: categories/Hexo/index.html
03:12:35.944 DEBUG Rendering HTML category: categories/Bash/index.html
03:12:35.994 DEBUG Rendering HTML category: categories/Github/index.html
03:12:36.054 DEBUG Rendering HTML category: categories/Nginx/index.html
03:12:36.116 DEBUG Rendering HTML index: index.html
03:12:36.193 DEBUG Rendering HTML index: page/2/index.html
03:12:36.261 DEBUG Rendering HTML tag: tags/Debian/index.html
03:12:36.312 DEBUG Rendering HTML tag: tags/TLS-1-3/index.html
03:12:36.377 DEBUG Rendering HTML tag: tags/NexT/index.html
03:12:36.431 DEBUG Rendering HTML tag: tags/Hexo/index.html
03:12:36.488 DEBUG Rendering HTML tag: tags/HTTPS/index.html
03:12:36.542 DEBUG Rendering HTML tag: tags/gulp/index.html
03:12:36.598 DEBUG Rendering HTML tag: tags/Bash/index.html
03:12:36.650 DEBUG Rendering HTML tag: tags/pixiv/index.html
03:12:36.713 DEBUG Rendering HTML tag: tags/Github/index.html
03:12:36.760 DEBUG Rendering HTML tag: tags/Nginx/index.html
03:12:36.809 DEBUG Rendering HTML tag: tags/444/index.html
03:12:37.077 DEBUG Rendering HTML tag: tags/Fail2ban/index.html
03:12:37.705 INFO  Generated: leancloud_counter_security_urls.json
03:12:37.706 INFO  Generated: atom.xml
03:12:37.707 INFO  Generated: search.xml
03:12:37.723 INFO  Generated: index.html
03:12:37.922 INFO  Generated: ding.mp3
03:12:37.989 INFO  Generated: robots.txt
03:12:38.079 INFO  Generated: leancloud.memo
03:12:38.116 INFO  Generated: favorites/index.html
03:12:38.201 INFO  Generated: categories/index.html
03:12:38.202 INFO  Generated: photos/index.html
03:12:38.203 INFO  Generated: tags/index.html
03:12:38.204 INFO  Generated: post/add-today-poetry-for-theme-next.html
03:12:38.205 INFO  Generated: archives/index.html
03:12:38.207 INFO  Generated: post/pixiv-manga-download-script.html
03:12:38.314 INFO  Generated: post/compiling-and-installing-nginx.html
03:12:38.473 INFO  Generated: post/block-malicious-access-to-nginx.html
03:12:38.587 INFO  Generated: post/use-gulp-to-minify-hexo.html
03:12:38.707 INFO  Generated: post/backup-files-to-github-private-repo.html
03:12:38.869 INFO  Generated: post/debian-install-haproxy.html
03:12:38.876 INFO  Generated: post/acme.sh-and-letsencrypt.html
03:12:39.013 INFO  Generated: image/algolia_logo.svg
03:12:39.090 INFO  Generated: tags/Fail2ban/index.html
03:12:39.092 INFO  Generated: images/algolia_logo.svg
03:12:39.295 INFO  Generated: image/android-chrome-192x192.png
03:12:39.408 INFO  Generated: image/alipay.png
03:12:39.409 INFO  Generated: image/android-chrome-256x256.png
03:12:39.501 INFO  Generated: image/apple-touch-icon.png
03:12:39.645 INFO  Generated: image/avatar.png
03:12:39.745 INFO  Generated: image/avatar.gif
03:12:39.747 INFO  Generated: image/bitcoin.png
03:12:39.748 INFO  Generated: image/browserconfig.xml
03:12:39.749 INFO  Generated: image/cc-by-nc-nd.svg
03:12:39.750 INFO  Generated: image/cc-by-nc-sa.svg
03:12:39.751 INFO  Generated: image/cc-by-nd.svg
03:12:39.751 INFO  Generated: image/cc-by-nc.svg
03:12:39.752 INFO  Generated: image/cc-by-sa.svg
03:12:39.753 INFO  Generated: image/cc-by.svg
03:12:39.754 INFO  Generated: image/favicon-hidden.png
03:12:39.755 INFO  Generated: image/cc-zero.svg
03:12:39.757 INFO  Generated: image/favicon-16x16.png
03:12:39.758 INFO  Generated: image/favicon-32x32.png
03:12:39.891 INFO  Generated: image/loading.gif
03:12:39.958 INFO  Generated: image/logo.svg
03:12:40.060 INFO  Generated: image/mstile-150x150.png
03:12:40.141 INFO  Generated: image/placeholder.gif
03:12:40.142 INFO  Generated: image/quote-l.svg
03:12:40.143 INFO  Generated: image/safari-pinned-tab.svg
03:12:40.218 INFO  Generated: image/quote-r.svg
03:12:40.320 INFO  Generated: image/show.ico
03:12:40.421 INFO  Generated: image/searchicon.png
03:12:40.518 INFO  Generated: image/wechatpay.png
03:12:40.519 INFO  Generated: image/site.webmanifest
03:12:40.610 INFO  Generated: images/avatar.gif
03:12:40.735 INFO  Generated: images/apple-touch-icon-next.png
03:12:40.736 INFO  Generated: images/cc-by-nc-sa.svg
03:12:40.737 INFO  Generated: images/cc-by-nd.svg
03:12:40.738 INFO  Generated: images/cc-by-nc-nd.svg
03:12:40.739 INFO  Generated: images/cc-by-nc.svg
03:12:40.740 INFO  Generated: images/cc-by-sa.svg
03:12:40.743 INFO  Generated: images/cc-by.svg
03:12:40.817 INFO  Generated: images/cc-zero.svg
03:12:40.818 INFO  Generated: images/favicon-32x32-next.png
03:12:40.819 INFO  Generated: images/favicon-16x16-next.png
03:12:40.820 INFO  Generated: images/favicon-hidden.png
03:12:40.821 INFO  Generated: images/logo.svg
03:12:40.822 INFO  Generated: images/quote-r.svg
03:12:40.823 INFO  Generated: images/quote-l.svg
03:12:40.830 INFO  Generated: archives/2019/01/index.html
03:12:40.983 INFO  Generated: page/2/index.html
03:12:41.092 INFO  Generated: archives/2019/02/index.html
03:12:41.093 INFO  Generated: categories/HAProxy/index.html
03:12:41.094 INFO  Generated: categories/NexT/index.html
03:12:41.095 INFO  Generated: archives/2019/03/index.html
03:12:41.096 INFO  Generated: categories/Hexo/index.html
03:12:41.097 INFO  Generated: categories/Bash/index.html
03:12:41.097 INFO  Generated: categories/Github/index.html
03:12:41.098 INFO  Generated: categories/Nginx/index.html
03:12:41.099 INFO  Generated: tags/NexT/index.html
03:12:41.100 INFO  Generated: tags/Debian/index.html
03:12:41.102 INFO  Generated: tags/gulp/index.html
03:12:41.103 INFO  Generated: tags/TLS-1-3/index.html
03:12:41.105 INFO  Generated: tags/Hexo/index.html
03:12:41.107 INFO  Generated: tags/pixiv/index.html
03:12:41.110 INFO  Generated: tags/Bash/index.html
03:12:41.251 INFO  Generated: tags/Github/index.html
03:12:41.257 INFO  Generated: tags/Nginx/index.html
03:12:41.258 INFO  Generated: tags/444/index.html
03:12:41.332 INFO  Generated: categories/HTTPS/index.html
03:12:41.333 INFO  Generated: tags/HTTPS/index.html
03:12:41.336 INFO  Generated: archives/2019/index.html
03:12:41.338 INFO  Generated: extra/fork-awesome/css/fork-awesome.css
03:12:41.339 INFO  Generated: css/nightmode.css
03:12:41.340 INFO  Generated: js/next-boot.js
03:12:41.341 INFO  Generated: image/feature/pixiv.gif
03:12:41.342 INFO  Generated: image/test/1.jpg
03:12:41.344 INFO  Generated: css/main.css
03:12:41.505 INFO  Generated: js/post-details.js
03:12:41.664 INFO  Generated: js/utils.js
03:12:41.787 INFO  Generated: js/scrollspy.js
03:12:41.799 INFO  Generated: extra/fork-awesome/css/fork-awesome.min.css.map
03:12:41.805 INFO  Generated: extra/fork-awesome/css/v5-compat.min.css.map
03:12:41.808 INFO  Generated: extra/fork-awesome/css/v5-compat.css
03:12:41.884 INFO  Generated: extra/fork-awesome/css/v5-compat.min.css
03:12:41.889 INFO  Generated: extra/fork-awesome/css/fork-awesome.min.css
03:12:41.890 INFO  Generated: extra/fork-awesome/fonts/forkawesome-webfont.woff2
03:12:41.891 INFO  Generated: extra/fork-awesome/fonts/forkawesome-webfont.ttf
03:12:41.970 INFO  Generated: extra/fork-awesome/fonts/forkawesome-webfont.woff
03:12:41.981 INFO  Generated: extra/fork-awesome/fonts/forkawesome-webfont.eot
03:12:41.985 INFO  Generated: js/schemes/muse.js
03:12:41.995 INFO  Generated: extra/fork-awesome/fonts/forkawesome-webfont.svg
03:12:42.083 INFO  Generated: image/feature/NexT.jpg
03:12:42.157 INFO  111 files generated in 9.28 s
03:12:42.222 DEBUG Database saved
03:12:42.264 WARN  [Hexo-Filters]---------------------------------before_exit starts
03:12:42.265 WARN  [Hexo-Events]---------------------------------exit starts
```

**Comment by 1v9** [comment]
@jiangtj I added hexo.log to every event and filter, that's easy to show process.
`config.ignore` is loaded very early.

**Comment by 1v9** [comment]
@stevenjoezhang @jiangtj Have a look?

File: scripts/filters/minify.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+hexo.extend.filter.register('after_generate', () => {
+  const theme = hexo.theme.config;
+  if (!theme.minify) return;
+
+  const lists = hexo.route.list();
+  const velocity = lists.filter(list => list.includes('lib/velocity'));
+  const fontawesome = lists.filter(list => list.includes('lib/font-awesome'));
+
+  if (!theme.motion.enable) {
```
Comment (stevenjoezhang) [inline scripts/filters/minify.js]: `if (!theme.motion.enable || (theme.vendors.velocity && theme.vendors.velocity_ui)) {`

Comment (1v9) [reply scripts/filters/minify.js]: fixed

**Pull request merged**